### PR TITLE
Utility to decode [SchemaProto] and [EntityTypeProto]

### PR DESCRIPTION
**Pull request by bgogul**
Only singletons and names are handled in this PR.  There are some TODOS in the code, which will be handled in subsequent PRs.

### CODE REVIEW

File: java/arcs/core/data/proto/SchemaProtoDecoder.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * Returns the fields in the [SchemaProto] as a Kotlin [SchemaFields] instance.
+ */
+fun SchemaProto.decodeFields(): SchemaFields {
+    val singletons = mutableMapOf<FieldName, FieldType>()
+    val collections = mutableMapOf<FieldName, FieldType>()
+    for ((name, type) in getFieldsMap()) {
+        when (type.getDataCase()) {
+            TypeProto.DataCase.PRIMITIVE -> singletons[name] = type.decodeAsFieldType()
+            // TODO: TypeProto.DataCase.COLLECTIONS
```
Comment (piotrswigon) [inline java/arcs/core/data/proto/SchemaProtoDecoder.kt]: + TODO references

File: java/arcs/core/data/proto/SchemaProtoDecoder.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                    "decodeFields for ${type.getDataCase().name} is not implemented.")
+        }
+    }
+    return SchemaFields(singletons, collections)
+}
+
+/**
+ * Converts a [SchemaProto] proto instance into a Kotlin [Schema] instance.
+ */
+fun SchemaProto.decode() = Schema(names = decodeNames(), fields = decodeFields(), hash = "")
```
Comment (piotrswigon) [inline java/arcs/core/data/proto/SchemaProtoDecoder.kt]: nice :)

File: java/arcs/core/data/proto/SchemaProtoDecoder.kt (hunk 1 above)
Comment (bgogul) [reply java/arcs/core/data/proto/SchemaProtoDecoder.kt]: Actually, `SchemaFields` has only two properties `singletons` and `collections`: https://github.com/PolymerLabs/arcs/blob/0f0f82a439cd00b2ea451f487bc3f96ff63635f1/java/arcs/core/data/SchemaFields.kt#L45-L48

What would a reference type map to? I believe it has to be a `singleton`?

Comment (piotrswigon) [reply java/arcs/core/data/proto/SchemaProtoDecoder.kt]: hmmm... schema can have a primitive field type which is a reference or a collection of references. I don't see how singleton relates to this.

Comment (bgogul) [reply java/arcs/core/data/proto/SchemaProtoDecoder.kt]: `Schema` class in Kotlin has a fields property, which is of type `SchemaFields` type shown avid. I am trying to determine how to translate the `fields` in `SchemaProto`:
```
message SchemaProto {
  repeated string names = 1;
  map<string, TypeProto> fields = 2;
}
```

Comment (piotrswigon) [reply java/arcs/core/data/proto/SchemaProtoDecoder.kt]: I just looked into SchemaFields, the reference over there is the EntityRef.

so if the type of the schema field in the proto is:
```
reference {
  referred_type {
    entity {
      schema {
       (...)
       hash: "abcd"
      }
    }
  }
}
```

it maps to `EntitiyRef("abcd")`

Comment (bgogul) [reply java/arcs/core/data/proto/SchemaProtoDecoder.kt]: Summarizing the offline discussion for the record. Collections of any type go into `collections`. Others go into `singletons`. We don't have support for tuples in Kotlin yet, and will be skipped for now.

**Pull request merged**